src/test/integration/mocks/LibProofGen.t.sol

```diff
@@ -0,0 +1,462 @@
+    /// like the validator tree, as each leaf is a validator's unique validatorFields.
+    /// However, for the balances tree, the leaves may not be distinct. To get around this,
+    /// BeaconChainMock._createValidator adds "dummy" validators every 4 validators created,
+    /// with a unique balance value. This ensures each balance root is unique.
+    function build(Tree storage tree, bytes32[] memory leaves, uint height) internal returns (bytes32 root) {
+        for (uint depth = 0; depth < height; depth++) {
+            uint newLength = (leaves.length + 1) / 2;
+            bytes32[] memory newLeaves = new bytes32[](newLength);
+
+            // Hash each pair of nodes in this level of the tree
+            for (uint i = 0; i < newLength; i++) {
+                uint leftIdx = 2 * i;
+                uint rightIdx = leftIdx + 1;
+
+                // Get left leaf
+                bytes32 leftLeaf = leaves[leftIdx];
+
+                // Calculate right leaf
+                bytes32 rightLeaf;
+                if (rightIdx < leaves.length) rightLeaf = leaves[rightIdx];
+                else rightLeaf = _getFillerNode(depth);
+
+                // Hash left and right
+                bytes32 result = sha256(abi.encodePacked(leftLeaf, rightLeaf));
+                newLeaves[i] = result;
+
+                // Record results, used to generate individual proofs later:
+                // Record left and right as siblings
+                tree.siblings[leftLeaf] = rightLeaf;
+                tree.siblings[rightLeaf] = leftLeaf;
+                // Record the result as the parent of left and right
+                tree.parents[leftLeaf] = result;
+                tree.parents[rightLeaf] = result;
+            }
+
+            // Move up one level
+            leaves = newLeaves;
+        }
+
+        require(leaves.length == 1, "LibProofGen.build: invalid tree somehow");
+        return leaves[0];
+    }
+
+    /// @dev Fetch unique filler node for given depth
+    function _getFillerNode(uint depth) internal view returns (bytes32) {
+        bytes32 curNode = Merkle.merkleizeSha256(new bytes32[](8));
+
+        for (uint i = 1; i < depth; i++) {
+            curNode = sha256(abi.encodePacked(curNode, curNode));
+        }
+
+        return curNode;
+    }
+
+    // /// @dev Pre-generate filler nodes used to fill in incomplete merkle trees at various depths
+    // function _calcZeroNodes() internal pure returns (bytes32[] memory) {
+    //     // Calculate nodes of empty merkle tree
+    //     bytes32 curNode = Merkle.merkleizeSha256(new bytes32[](8));
+    //     bytes32[] memory zeroNodes = new bytes32[](MAX_TREE_HEIGHT);
+    //     zeroNodes[0] = curNode;
+
+    //     for (uint i = 1; i < zeroNodes.length; i++) {
+    //         zeroNodes[i] = sha256(abi.encodePacked(curNode, curNode));
+    //         curNode = zeroNodes[i];
+    //     }
+
+    //     return zeroNodes;
```

Comment:
Note to self - remove

Reply:
[179e268](https://github.com/Layr-Labs/eigenlayer-contracts/pull/1363/commits/179e268ddbff1cfec143220464e73b74354888c8)